### PR TITLE
Frontend: Migrate to React@17

### PR DESCRIPTION
**Pull request by BrianCraig**
This is a React Migration that we been working on with @thisconnect.

Main points are:
- Migrated from `preact@8` to `react@17`.  
- Updated i18n library, updating custom logic, HOCs, and migrated decorators.  
- Migrated existing tests to `enzyme` and `@testing-library/react`, use the last one for new test suites.  
- renamed css modules to `.module.css`.  
- changed from named imports to default imports on css modules.  
- migrated `Ref functions` to `Ref objects`, and components which implement it.  
- migrated to `react-router-dom`.  
- Decoupled Routes from `App` into `AppRouter`.  
- implemented `route` fn as in `preact-router`.  
- migrated from `yarn` to `npm`.  

Benefits:  
- Able to use Hooks, Context, Suspense, and more tools from `react`.  
- Able to use the React developer tools.  
- more libraries able to be integrated.  

PTAL @benma @x1ddos

### CODE REVIEW

**Comment by BrianCraig** [comment]
Blocked by #1530

File: frontends/web/src/components/icon/assets/link-white.svg [hunk 1]
```diff
@@ -1,31 +1,3 @@
...
-    </rdf:RDF>
-  </metadata>
-  <defs
-     id="defs8" />
-  <path
-     id="path4599"
-     d="m 16.032206,0.03131126 c -8.8370023,0 -16.00000254,7.16300004 -16.00000254,16.00000074 0,8.837 7.16300024,16 16.00000254,16 8.837,0 16,-7.163 16,-16 0,-8.8370007 -7.163,-16.00000074 -16,-16.00000074 z m -0.02734,4.36914104 2.201172,1.265624 6,3.4667975 2.201172,1.2675792 v 12 l -2.201172,1.265624 -5.945312,3.466797 -2.201172,1.267579 -2.201172,-1.267579 -6.0546899,-3.466797 -2.203125,-1.265624 v -12 L 7.8036921,9.1328738 13.801741,5.6660763 Z m 0,5.0644535 -6.000002,3.4667972 v 6.935546 l 6.000002,3.46875 6,-3.46875 v -6.935546 z"
-     style="fill:#777777;fill-opacity:1;fill-rule:evenodd" />
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32">
+  <path fill="#777" fill-rule="evenodd" d="M16.0322.0313c-8.837 0-16 7.163-16 16s7.163 16 16 16 16-7.163 16-16-7.163-16-16-16zm-.0273 4.3692L18.206 5.666l6 3.4668 2.2012 1.2676v12L24.206 23.666l-5.9453 3.4668-2.2011 1.2676-2.2012-1.2676-6.0547-3.4668-2.2031-1.2656v-12l2.203-1.2676 5.9981-3.4668Zm0 5.0644-6 3.4668v6.9355l6 3.4688 6-3.4688v-6.9355z"/>
```
Comment (thisconnect) [inline frontends/web/src/components/icon/assets/link-white.svg:2]: Sound a bug:

Screen Shot 2021-12-16 at 11 00 07 AM

File: frontends/web/src/components/icon/logo.tsx [hunk 2]
```diff
@@ -67,7 +66,7 @@ export const AppLogoInverted = (props: GenericProps) => <img {...props} draggabl
 export const BitBoxSwiss = (props: GenericProps) => <img {...props} draggable={false} src={BitBoxSwissLogo} alt="BitBox" className={style.logo} />;
 export const BitBoxSwissInverted = (props: GenericProps) => <img {...props} draggable={false} src={BitBoxSwissInvertedLogo} alt="BitBox" className={style.logo} />;
 export const Shift = (props: GenericProps) => <img {...props} draggable={false} src={ShiftLogo} alt="Shift Crypto" className={style.logo} />;
-export const SwissMadeOpenSource = (props: GenericProps) => <img {...props} draggable={false} src={SwissOpenSource} alt="Swiss Made Open Source" className={[style.swissOpenSource, props.large ? style.large : ''].join(' ')} />;
+export const SwissMadeOpenSource = ({large: boolean, ...props}: GenericProps) => <img {...props} draggable={false} src={SwissOpenSource} alt="Swiss Made Open Source" className={[style.swissOpenSource, props.large ? style.large : ''].join(' ')} />;
```
Comment (thisconnect) [inline frontends/web/src/components/icon/logo.tsx:69]: @BrianCraig `large` is deconstructed but then there is `props.large` in `className={[style.swissOpenSource, props.large ? style.large : '']`

File: frontends/web/src/components/language/language.tsx [hunk 3]
```diff
@@ -66,17 +65,9 @@ const defaultLanguages = [
...
         super(props);
         const languages = this.props.languages || defaultLanguages;
-        if (extraLanguages) {
-            extraLanguages.split(',').forEach(code => {
-                languages.push({
-                    code,
-                    display: code,
-                } as TLanguage);
-            });
-        }
```
Comment (thisconnect) [inline frontends/web/src/components/language/language.tsx:79]: Why did we remove this again?

was extraLanguages only used for testing maybe?

File: frontends/web/src/components/language/language.tsx [hunk 4]
```diff
@@ -220,6 +185,6 @@ class LanguageSwitch extends Component<Props, State> {
     }
 }
 
-const TranslatedLanguageSwitcher = translateDecorator<LanguageSwitchProps>()(LanguageSwitch);
+const TranslatedLanguageSwitcher = withTranslation()(LanguageSwitch);
 
 export { TranslatedLanguageSwitcher as LanguageSwitch };
```
Comment (thisconnect) [inline frontends/web/src/components/language/language.tsx:190]: Test: language switch and fallback to system default

File: frontends/web/src/components/message/message.tsx [hunk 5]
```diff
@@ -14,21 +14,21 @@
...
     type?: 'message' | 'success' | 'info' | 'warning' | 'error';
-    style?: string;
+    style?: React.CSSProperties;
 }
 
 export function Message({
     hidden,
     type = 'message',
-    style = '',
+    style,
```
Comment (thisconnect) [inline frontends/web/src/components/message/message.tsx:29]: style is not used anymore in the codebase, except in a test, we could remove or pass with `...props`

File: frontends/web/src/components/password.jsx [hunk 6]
```diff
@@ -321,5 +323,6 @@ function hasCaps({ key }) {
     }
     // ideally we return event.getModifierState('CapsLock')) but this currently does always return false in Qt
     // @ts-ignore (event can be undefined and shiftKey exists only on MouseEvent but not Event)
+    // eslint-disable-next-line no-restricted-globals
     return key.toUpperCase() === key && key.toLowerCase() !== key && !event.shiftKey;
 }
```
Comment (thisconnect) [inline frontends/web/src/components/password.jsx:328]: Test BB01 password and setup password

File: frontends/web/src/components/sidebar/sidebar.tsx [hunk 7]
```diff
@@ -290,5 +284,5 @@ const subscribeHOC = subscribe<SubscribedProps, SharedPanelProps & SidebarProps
 )(Sidebar);
 
 const guideShareHOC = share<SharedPanelProps, SidebarProps & TranslateProps>(panelStore)(subscribeHOC);
-const translateHOC = translate<SidebarProps>()(guideShareHOC);
+const translateHOC = translate()(guideShareHOC);
 export { translateHOC as Sidebar };
```
Comment (thisconnect) [inline frontends/web/src/components/sidebar/sidebar.tsx:288]: Test sidebar links and navigation

File: frontends/web/src/components/transactions/transaction.tsx [hunk 8]
```diff
@@ -395,6 +387,6 @@ class Transaction extends Component<Props, State> {
     }
 }
 
-const HOC = translate<TransactionProps>()(Transaction);
+const HOC = translate()(Transaction);
 
 export { HOC as Transaction };
```
Comment (thisconnect) [inline frontends/web/src/components/transactions/transaction.tsx:392]: Test transaction notes and details

File: frontends/web/src/components/wait-dialog/wait-dialog.tsx [hunk 9]
```diff
@@ -170,5 +163,5 @@ class WaitDialog extends Component<Props, State> {
     }
 }
 
-const TranslatedWaitDialog = translate<WaitDialogProps>()(WaitDialog);
+const TranslatedWaitDialog = translate()(WaitDialog);
 export { TranslatedWaitDialog as WaitDialog };
```
Comment (thisconnect) [inline frontends/web/src/components/wait-dialog/wait-dialog.tsx:167]: Test wait-dialog (the blocking dialog that sometimes appears) i.e. when sending

File: frontends/web/src/routes/account/account.tsx [hunk 10]
```diff
@@ -417,5 +419,5 @@ const loadHOC = load<LoadedAccountProps, AccountProps & TranslateProps>(({ code
     config: 'config',
 }))(Account);
 
-const HOC = translate<AccountProps>()(loadHOC);
+const HOC = translate()(loadHOC);
 export { HOC as Account };
```
Comment (thisconnect) [inline frontends/web/src/routes/account/account.tsx:423]: Test buttons on account overview

Comment (thisconnect) [reply frontends/web/src/routes/account/account.tsx:423]: Bug: for some reason I the account-info link is gone 🤔

File: frontends/web/src/routes/account/add/add.tsx [hunk 11]
```diff
@@ -289,6 +289,6 @@ class AddAccount extends Component<Props, State> {
     }
 }
 
-const HOC = translate<AddAccountProps>()(AddAccount);
+const HOC = translate()(AddAccount);
 
 export { HOC as AddAccount };
```
Comment (thisconnect) [inline frontends/web/src/routes/account/add/add.tsx:294]: Test: add new account

File: frontends/web/src/routes/account/send/send.tsx [hunk 12]
```diff
@@ -462,17 +466,17 @@ class Send extends Component<Props, State> {
 
     private toggleCoinControl = () => {
         this.setState(({ activeCoinControl }) => {
-            if (activeCoinControl && this.utxos) {
-                (this.utxos as any).getWrappedInstance().clear();
+            if (activeCoinControl && this.utxos.current) {
+                try {
+                    (this.utxos.current as any).getWrappedInstance().clear();
```
Comment (thisconnect) [inline frontends/web/src/routes/account/send/send.tsx:471]: I think we should get rid of this.

File: frontends/web/src/routes/account/send/send.tsx [hunk 13]
```diff
@@ -878,5 +891,5 @@ class Send extends Component<Props, State> {
     }
 }
 
-const TranslatedSend = translate<SendProps>()(Send);
+const TranslatedSend = translate()(Send);
 export { TranslatedSend as Send };
```
Comment (thisconnect) [inline frontends/web/src/routes/account/send/send.tsx:895]: Test: send with coin-control

File: frontends/web/src/routes/account/summary/chart.tsx [hunk 14]
```diff
@@ -443,7 +443,7 @@ class Chart extends Component<Props, State> {
     }
 }
 
-const HOC = translate<ChartProps>()(Chart);
+const HOC = translate()(Chart);
 
 export { HOC as Chart };
```
Comment (thisconnect) [inline frontends/web/src/routes/account/summary/chart.tsx:449]: Test chart tooltip

File: frontends/web/src/routes/buy/moonpay.tsx [hunk 15]
```diff
@@ -120,5 +123,5 @@ class Moonpay extends Component<Props, State> {
 const loadHOC = load<LoadedBuyProps, BuyProps & TranslateProps>(({ code }) => ({
     moonpay: `exchange/moonpay/buy/${code}`,
 }))(Moonpay);
-const HOC = translate<BuyProps>()(loadHOC);
+const HOC = translate()(loadHOC);
 export { HOC as Moonpay };
```
Comment (thisconnect) [inline frontends/web/src/routes/buy/moonpay.tsx:127]: Test that moonpay iframe loads

File: frontends/web/src/routes/device/bitbox01/backups.tsx [hunk 16]
```diff
@@ -188,5 +187,5 @@ class Backups extends Component<Props, State> {
     }
 }
 
-const TranslatedBackups = translate<BackupsProps>()(Backups);
+const TranslatedBackups = translate()(Backups);
 export { TranslatedBackups as Backups };
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox01/backups.tsx:191]: Bonus test: BB01 long Backuplists

File: frontends/web/src/routes/device/bitbox01/create.jsx [hunk 17]
```diff
@@ -121,4 +121,4 @@ class Create extends Component {
     }
 }
 
-export default translate()(Create);
+export default withTranslation()(Create);
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox01/create.jsx:124]: Test BB01 create check that the backup name doesn't autocomplete

File: frontends/web/src/routes/device/bitbox01/settings/settings.jsx [hunk 18]
```diff
@@ -228,4 +232,4 @@ class Settings extends Component {
     }
 }
 
-export default translate()(Settings);
+export default withTranslation()(Settings);
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox01/settings/settings.jsx:235]: Test BB01 manage backups in the settings page

File: frontends/web/src/routes/device/bitbox02/components/password-entry/password-entry.tsx [hunk 19]
```diff
@@ -42,7 +48,7 @@ export function PasswordEntry({ children }: RenderableProps<IPasswordEntryProps>
                 // i.e. when re-plugin the BitBox the video doesn't play anymore
                 // https://github.com/preactjs/preact/issues/747#issuecomment-370905360
                 // looks like this can be removed with Preact10/React
-                ref={ref => replayVideo(ref)}
+                ref={ref}
                 className={styles.passwordGestures}
                 loop
                 muted
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox02/components/password-entry/password-entry.tsx:54]: Test unlock video

File: frontends/web/src/routes/device/bitbox02/components/password-entry/password-entry.tsx [hunk 20]
```diff
@@ -42,7 +48,7 @@ export function PasswordEntry({ children }: RenderableProps<IPasswordEntryProps>
                 // i.e. when re-plugin the BitBox the video doesn't play anymore
                 // https://github.com/preactjs/preact/issues/747#issuecomment-370905360
                 // looks like this can be removed with Preact10/React
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox02/components/password-entry/password-entry.tsx:50]: could remove the comment

File: frontends/web/src/routes/device/bitbox02/sdcardcheck.tsx [hunk 21]
```diff
@@ -85,5 +89,5 @@ class SDCardCheck extends Component<Props, State> {
 
 }
 
-const HOC = translate<SDCardCheckProps>()(SDCardCheck);
+const HOC = translate()(SDCardCheck);
 export { HOC as SDCardCheck };
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox02/sdcardcheck.tsx:93]: Test manage backups with and without SD card

File: frontends/web/src/routes/device/bitbox02/setdevicename.jsx [hunk 22]
```diff
@@ -124,4 +125,4 @@ class SetDeviceNameClass extends Component {
     }
 }
 
-export const SetDeviceName = translate()(SetDeviceNameClass);
+export const SetDeviceName = withTranslation()(SetDeviceNameClass);
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox02/setdevicename.jsx:128]: Test set device name

File: frontends/web/src/routes/device/bitbox02/upgradebutton.tsx [hunk 23]
```diff
@@ -113,5 +119,5 @@ class UpgradeButton extends Component<Props, State> {
     }
 }
 
-const HOC = translate<UpgradeButtonProps>()(UpgradeButton);
+const HOC = translate()(UpgradeButton);
 export { HOC as UpgradeButton };
```
Comment (thisconnect) [inline frontends/web/src/routes/device/bitbox02/upgradebutton.tsx:123]: Test upgrade firmware

File: frontends/web/src/routes/settings/settings.tsx [hunk 24]
```diff
@@ -169,7 +166,7 @@ class Settings extends Component<Props, State> {
                     <Header title={<h2>{t('settings.title')}</h2>}>
                         {
                             !deviceIDs.length && (
-                                <Link onClick={this.backHome} className="flex flex-row flex-items-center">
+                                <Link to='/' className="flex flex-row flex-items-center">
```
Comment (thisconnect) [inline frontends/web/src/routes/settings/settings.tsx:169]: Bug: if no devices are connected, there should be a home button in the header on the settings page but this isn't rendered. Gutfeeling: this is the same issue as why account-overview doesn't show account-info link.

@BrianCraig could you take a look?

File: frontends/web/src/utils/simplemarkup.test.tsx [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+    });
+    // it('should but doesnt support multiple strong elements', () => {
+    //     const span = shallow(<SimpleMarkup tagName="span" markup="<strong>foo</strong> <strong>bar</strong> <strong>baz</strong>" />);
+    //     expect(span.output().nodeName).toEqual('span');
+    //     expect(span.text()).toEqual('<strong>foo</strong> <strong>bar</strong> baz');
+    //     // Only the last strong element is currently supported :/
+    //     expect(span.output().children[0]).toEqual('<strong>foo</strong> <strong>bar</strong> ');
+    //     expect(span.output().children[1]).toEqual(<strong>baz</strong>)
+    // });
+});
```
Comment (thisconnect) [inline frontends/web/src/utils/simplemarkup.test.tsx:39]: I believe we can remove this test in favor or `markup.test.tsx`

File: frontends/web/tsconfig.json [hunk 26]
```diff
@@ -1,27 +1,31 @@
...
-    ]
+  "compilerOptions": {
+    "target": "es2017",
+    "lib": [
+      "dom",
+      "dom.iterable",
+      "esnext"
+    ],
+    "allowJs": true,
+    "noImplicitAny": false,
```
Comment (thisconnect) [inline frontends/web/tsconfig.json:10]: TODO: set to true

**Review by thisconnect — COMMENTED** [review]
I tested a bit and made comments where there is code changes that should be tested. Comments starting with "Test" could be additionally tested when we test the RC.

Also found 2 bugs, one is some SVG do not render, and it seems children (account-info link and home link) of the Header component doesn't render.

File: frontends/web/src/components/copy/Copy.tsx [hunk 27]
```diff
@@ -116,5 +112,5 @@ class CopyableInput extends Component<Props, State> {
     }
 }
 
-const TranslatedCopyableInput = translate<CopyableInputProps>()(CopyableInput);
+const TranslatedCopyableInput = translate()(CopyableInput);
 export { TranslatedCopyableInput as CopyableInput };
```
[the remark this answers is not in the document]
Comment (thisconnect) [reply frontends/web/src/components/copy/Copy.tsx:116]: Here interesting would be testing long ETH address that breaks onto 2 lines on small screen (android)

File: frontends/web/src/components/icon/logo.tsx (hunk 2 above)
Comment (BrianCraig) [reply frontends/web/src/components/icon/logo.tsx:69]: This is a problem that `large` is a `` prop, but interferes (not a boolean), and it was passed down to the img tag, when it's a component prop

File: frontends/web/src/components/language/language.tsx (hunk 3 above)
Comment (BrianCraig) [reply frontends/web/src/components/language/language.tsx:79]: This was removed on 29a0764ea4a5d4f412ec928055439a083c47d99d

File: frontends/web/src/components/message/message.tsx (hunk 5 above)
Comment (BrianCraig) [reply frontends/web/src/components/message/message.tsx:29]: I made an issue #1535

File: frontends/web/src/routes/account/account.tsx (hunk 10 above)
Comment (BrianCraig) [reply frontends/web/src/routes/account/account.tsx:423]: Do you have more info on this bug so i can check it? (how to reproduce it, expected result)

File: frontends/web/src/routes/device/bitbox02/components/password-entry/password-entry.tsx (hunk 20 above)
Comment (BrianCraig) [reply frontends/web/src/routes/device/bitbox02/components/password-entry/password-entry.tsx:50]: #1538

File: frontends/web/src/routes/settings/settings.tsx (hunk 24 above)
Comment (BrianCraig) [reply frontends/web/src/routes/settings/settings.tsx:169]: #1539

File: frontends/web/src/routes/account/account.tsx (hunk 10 above)
Comment (thisconnect) [reply frontends/web/src/routes/account/account.tsx:423]: Yes sorry, on account overview there should be a account info link in the header, with useful information such as xpub or coin type.

The link is passed as child to the header component, for some reason this doesn't render anymore.

Also probably this shouldn't be a `<a href>` I think, we should always route programatically.

https://github.com/digitalbitbox/bitbox-wallet-app/blob/ad414e3dad2e8926c90660fad8bb26a16affa772/frontends/web/src/routes/account/account.tsx#L307-L310

But that it doesn't render at all is maybe due to `hasChildren` here https://github.com/digitalbitbox/bitbox-wallet-app/blob/ad414e3dad2e8926c90660fad8bb26a16affa772/frontends/web/src/components/layout/header.tsx#L42

File: frontends/web/src/components/layout/version.tsx [hunk 28]
```diff
@@ -14,19 +15,22 @@
...
+    const [version, setVersion] = useState<string>('');
 
-type Props = VersionProps & TranslateProps;
+    useEffect(() => {
+        getVersion().then(setVersion);
+    }, []);
 
-function Version({ t, version }: PropsWithChildren<Props>) {
+    if (!version) {
+        return null;
```
Comment (thisconnect) [inline frontends/web/src/components/layout/version.tsx:31]: We could render a placeholder here.

**Review by thisconnect — APPROVED** [review]
LGTM

**Review by benma — APPROVED** [review]
nice! i did a lightweight test on linux and android.

the only thing i noticed that there is a whitescreen saying 'loading' for a splti second before the rest of the app loads.

**Pull request merged**